### PR TITLE
First push of hostpath-provisioner-operator.

### PR DESCRIPTION
**Pull request by awels**
First push of operator, *warning* rough around edges.

Signed-off-by: Alexander Wels <awels@redhat.com>

### CODE REVIEW

File: deploy/cluster_role.yaml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - pods
+  - persistentvolumeclaims
+  - persistentvolumes
+  - events
+  - configmaps
+  - secrets
```
Comment (mhenriks) [inline deploy/cluster_role.yaml]: What secrets we talking about here?

File: pkg/controller/hostpathprovisioner/contoller_test.go [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-name",
+				Namespace: "test-namespace",
+			},
+			Spec: v1alpha1.HostPathProvisionerSpec{
+				ImageRegistry:   "test-registry",
+				ImagePullPolicy: corev1.PullAlways,
+				ImageTag:        "canary",
+				PathConfig: v1alpha1.PathConfig{
+					Path:            "/tmp/test",
```
Comment (mhenriks) [inline pkg/controller/hostpathprovisioner/contoller_test.go]: Should we test that a CS with empty path is rejected?

File: deploy/cluster_role.yaml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+- apiGroups:
+  - hostpathprovisioner.kubevirt.io
+  resources:
+  - '*'
+  verbs:
+  - '*'
+- apiGroups:
+  - security.openshift.io
+  resources:
+  - securitycontextconstraints
```
[the remark this answers is not in the document]
Comment (awels) [reply deploy/cluster_role.yaml]: So there are 6 verbs right? create/update/delete/list/watch/get.
I need all of those to fully manage, so instead of spelling them all out, I used *

File: deploy/cluster_role.yaml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  - persistentvolumes
+  - events
+  - configmaps
+  - secrets
+  verbs:
+  - '*'
+- apiGroups:
+  - rbac.authorization.k8s.io
+  resources:
+  - clusterrolebindings
```
[the remark this answers is not in the document]
Comment (awels) [reply deploy/cluster_role.yaml]: But I need watch/list/get as well.

File: deploy/crds/hostpathprovisioner_cr.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+kind: HostPathProvisioner
+metadata:
+  name: hostpath-provisioner
+spec:
+  imagePullPolicy: IfNotPresent
+  imageRegistry: kubevirt
+  imageTag: latest
+  pathConfig:
+    path: "/var/hpvolumes"
+    useNamingPrefix: "false"
```
[the remark this answers is not in the document]
Comment (awels) [reply deploy/crds/hostpathprovisioner_cr.yaml]: Fixed

File: pkg/controller/hostpathprovisioner/contoller_test.go (hunk 2 above)
Comment (awels) [reply pkg/controller/hostpathprovisioner/contoller_test.go]: We would need a functional test for that, which would require a cluster to be up. I didn't want want to involve clusters at this point.

File: deploy/cluster_role.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply deploy/cluster_role.yaml]: yeah what I'm suggesting is to have multiple policies for `securitycontextconstraints`.  Get/List for any resource.  But restrict create/updates/delete to specific resourcenames

File: pkg/controller/hostpathprovisioner/contoller_test.go (hunk 2 above)
Comment (mhenriks) [reply pkg/controller/hostpathprovisioner/contoller_test.go]: What will the controller do if it encounters a CR with the path unset?

Comment (awels) [reply pkg/controller/hostpathprovisioner/contoller_test.go]: It will fail according to this: https://github.com/kubevirt/hostpath-provisioner/blob/master/cmd/provisioner/hostpath-provisioner.go#L74

File: pkg/apis/addtoscheme_hostpathprovisioner_v1alpha1.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	//secv1 "github.com/openshift/api/security/v1"
+	"kubevirt.io/hostpath-provisioner-operator/pkg/apis/hostpathprovisioner/v1alpha1"
+)
+
+func init() {
+	// Register the types with the Scheme so the components can map objects to GroupVersionKinds and back
+	AddToSchemes = append(AddToSchemes, v1alpha1.SchemeBuilder.AddToScheme)
+	// Register SCC types.
+	//AddToSchemes = append(AddToSchemes, secv1.SchemeBuilder.AddToScheme)
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/apis/addtoscheme_hostpathprovisioner_v1alpha1.go]: Fixed

File: deploy/cluster_role.yaml (hunk 1 above)
Comment (awels) [reply deploy/cluster_role.yaml]: No secrets, so I removed it.

File: deploy/cluster_role.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (awels) [reply deploy/cluster_role.yaml]: Gotcha, missed that distinction, I can do similar for a bunch of other resources as well. Let me work on updating that.

File: README.md [hunk 7]
```diff
@@ -1 +1,36 @@
...
+kind: HostPathProvisioner
+metadata:
+  name: hostpath-provisioner
+spec:
+  imagePullPolicy: IfNotPresent
+  imageRegistry: kubevirt/hostpath-provisioner #Registry to get the hostpath provisioner container from
+  imageTag: latest #Tag of the hostpath provisioner container
+  pathConfig:
+    path: "/var/hpvolumes" #The path of the directory on the node
+    useNamingPrefix: "false" #Make the PV name part of the directory that is created.
```
Comment (aglitke) [inline README.md]: Is it PV or PVC?  Just confirming...

File: README.md [hunk 8]
```diff
@@ -1 +1,36 @@
...
+```
+
+Note the yamls assume you want to deploy into the hostpath_provisioner namespace, which has to be created ahead of time. If you want to install into a different namespace you will have modify the ClusterRoleBinding namespace to match. The rest can be installed using -n to specify the namespace.
+
+## Deployment in OpenShift
+The operator will create the appropriate SecurityContextConstraints for the hostpath provisioner to work and assign the ServiceAccount to that SCC. However on each node you will have to give the directory you specify in the CR the appropriate selinx rules by running the following (assuming you pick /var/hpvolumes as your PathConfig path):
+```bash
+$ sudo chcon -R unconfined_u:object_r:svirt_sandbox_file_t:s0 /var/hpvolumes
+```
```
Comment (aglitke) [inline README.md]: Is this Openshift specific or distro specific?

File: deploy/crds/hostpathprovisioner_cr.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+apiVersion: hostpathprovisioner.kubevirt.io/v1alpha1
+kind: HostPathProvisioner
+metadata:
+  name: hostpath-provisioner
+spec:
+  imagePullPolicy: IfNotPresent
+  imageRegistry: quay.io/kubevirt
```
Comment (aglitke) [inline deploy/crds/hostpathprovisioner_cr.yaml]: We decided to host on quay?

File: pkg/controller/hostpathprovisioner/daemonset.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	hostpathprovisionerv1alpha1 "kubevirt.io/hostpath-provisioner-operator/pkg/apis/hostpathprovisioner/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// TODO make it so you can pass the image when creating the operator CRD.
+	provisionerImage = "hostpath-provisioner"
```
Comment (mhenriks) [inline pkg/controller/hostpathprovisioner/daemonset.go]: Don't think this is the same upstream/downstream for other components.  So may have to have it be a env var.  Env var is prob preferable to CR property in order to be consistent with other components.

File: README.md (hunk 7 above)
Comment (awels) [reply README.md]: PVC, you are right, I will fix.

File: deploy/crds/hostpathprovisioner_cr.yaml (hunk 9 above)
Comment (awels) [reply deploy/crds/hostpathprovisioner_cr.yaml]: I talked to Fabian, and we are doing all new things on quay.

File: README.md (hunk 8 above)
Comment (awels) [reply README.md]: This actually has to happen all the time, I should not have put it under Openshift heading.

File: pkg/controller/hostpathprovisioner/daemonset.go (hunk 10 above)
Comment (awels) [reply pkg/controller/hostpathprovisioner/daemonset.go]: Good point, let me see how other components solve this.

**Comment by awels** [comment]
I realize the pattern for all the different objects in the reconcile loop is the same, and I could probably make a generic reconcile loop similar to CDI, but I think currently that is overkill for this operator.

File: pkg/controller/hostpathprovisioner/scc.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	return r.client.Delete(context.TODO(), scc)
+}
+
+func (r *ReconcileHostPathProvisioner) addSCCFinalizer(reqLogger logr.Logger, cr *hostpathprovisionerv1alpha1.HostPathProvisioner) error {
+	if used, err := r.checkSCCUsed(); used == false {
+		return err
+	}
+	if len(cr.GetFinalizers()) < 1 && cr.GetDeletionTimestamp() == nil {
+		reqLogger.Info("Adding SCC Finalizer")
+		cr.SetFinalizers([]string{"finalizer.scc.hostpath-provisioner"})
```
Comment (mhenriks) [inline pkg/controller/hostpathprovisioner/scc.go]: Why is this needed?

Comment (awels) [reply pkg/controller/hostpathprovisioner/scc.go]: So even though the operator controller owns the scc, when the CR is removed, it didn't remove the scc automatically as well. I have a feeling it is an Open Shift protection thing. So I had to use a finalizer to specifically remove the SCC when the CR is removed.
